[RNMobile] Refactor: gallery to nested image blocks

dratwas opened the pull request:
<!-- Learn the overall process and best practices for pull requests at https://github.com/WordPress/gutenberg/blob/master/docs/contributors/repository-management.md#pull-requests. -->

## Description
Refactors the Gallery block to be a wrapper around Image blocks using InnerBlocks. This is a mobile implementation of https://github.com/WordPress/gutenberg/pull/25940

### What is done:
In order to make it works, I had to do a lot of changes. The main change is related to the "Tile" component. Till now the Gallery block used the Tile component to set the correct layout of the gallery, however, now we use inner blocks for rendering and there is no possibility to use it in the same way since the InnerBlock component is responsible for the layout. I moved the code from gallery/tile to the block-list/grid-item and use it there. In order to make it works, I also had to add the `gridProperties` prop to set the correct number of columns (there could be more properties in the future).

In the base PR the author moved some code from media-placeholder to the gallery/edit and I had also done the same to be compatible with that.

Another important thing is that the image block is not prepared to be used in the galley. I mean that images in the gallery have the same height etc so I had also make some changes in the image block. I added a prop `isGallery` and` imageCrop` to determine how the image should look like.

I couldn't use the `useParentAttributes` to get inherited attributes from the gallery because the image/edit.native.js is a class. I implemented it in the class way (using the componentDidUpdate) however, it will be probably changed by the web team and should be implemented here as well - I describe it below.

I also had to add some checks in the `link-settings/index.native.js` to prevent setting the URL when it's not necessary because it was setting the wrong URL when using the nested URL from the gallery. 

I implemented the possibility to use the `blockType.usesContext` as it is done on the web. It was used to pass down some attributes from the gallery to the image.

The thing that is still under development and I believe we should wait for the final implementation is how the inner blocks should behave when we change some attributes in the gallery. There are 3 attributes that are in the gallery and also in the image block:
- inkDestination
- linkTarget
- sizeSlug

The discussion about that https://github.com/WordPress/gutenberg/pull/25940#discussion_r574300020

At this moment they are passed down through the context (that's why I had to add the possibility to use the BlockContext), however, there is still discussion on how it should be done and how the inner blocks should act when we change one of the listed attributes. 

I really suggest waiting for the final decision since I had to change the implementation a few times in the meantime because of the new review.

### What should be done:

First of all, as I mentioned we should wait till the web team decides about all implementation details because I don't think that it makes sense to implement it parallelly since they change his approach quite frequently.

Then this PR should be rebased to the latest `refactor/gallery-to-nested-image-blocks`

Then depending on the web team's decision about how to handle the inherited attributes that are in the gallery block and the image block as well, made changes in the image/edit.native.js.

The important thing is also that the web team wants to keep the old implementation and new implementation together:
https://github.com/WordPress/gutenberg/pull/25940#issuecomment-777256259


## How has this been tested?
All sanity tests for gallery 
Gallery-1

- [ ] Gallery block - Close/Re-open post with an ongoing image upload - [steps](https://github.com/wordpress-mobile/test-cases/blob/master/test-cases/gutenberg/gallery.md#tc001)
- [ ] Gallery block - Close post with an ongoing image upload - [steps](https://github.com/wordpress-mobile/test-cases/blob/master/test-cases/gutenberg/gallery.md#tc002)

Gallery-2

- [ ] Gallery block - Insert image from device (failing) - [steps](https://github.com/wordpress-mobile/test-cases/blob/master/test-cases/gutenberg/gallery.md#tc006)
- [ ] Gallery block - Insert image from device (cancel) - [steps](https://github.com/wordpress-mobile/test-cases/blob/master/test-cases/gutenberg/gallery.md#tc009)

Gallery-3

- [ ] Gallery block - Add Gallery Caption - [steps](https://github.com/wordpress-mobile/test-cases/blob/master/test-cases/gutenberg/gallery.md#tc003)
- [ ] Gallery block - Add Gallery Image Caption - [steps](https://github.com/wordpress-mobile/test-cases/blob/master/test-cases/gutenberg/gallery.md#tc004)
- [ ] Gallery block - Try adding same images from WP Media library and moving the images around - [steps](https://github.com/wordpress-mobile/test-cases/blob/master/test-cases/gutenberg/gallery.md#tc010)

Gallery-4

- [ ] Gallery block - Settings: Link to - [steps](https://github.com/wordpress-mobile/test-cases/blob/master/test-cases/gutenberg/gallery.md#tc012)
- [ ] Gallery block - Settings: Column number - [steps](https://github.com/wordpress-mobile/test-cases/blob/master/test-cases/gutenberg/gallery.md#tc013)
- [ ] Gallery block - Settings: Crop images - [steps](https://github.com/wordpress-mobile/test-cases/blob/master/test-cases/gutenberg/gallery.md#tc014)


## Screenshots <!-- if applicable -->

![ezgif com-gif-maker](https://user-images.githubusercontent.com/16336501/99065160-4148fa80-25a7-11eb-9849-f0a1212274ae.gif)

## Types of changes
New feature (non-breaking change which adds functionality)
Refactors the Gallery block to be a wrapper around Image blocks using InnerBlocks. 

## Checklist:
- [ ] My code is tested.
- [ ] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/ -->
- [ ] My code follows the accessibility standards. <!-- Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/accessibility/ -->
- [ ] My code has proper inline documentation. <!-- Guidelines: https://developer.wordpress.org/coding-standards/inline-documentation-standards/javascript/ -->
- [ ] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR. <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/master/docs/contributors/native-mobile.md -->

Commit:
Refactor gallery to use nested image blocks

Commit:
Fix issue with images not loading on first selection from media gallery

Commit:
Fix issue with images not loading on first selection from media gallery

Commit:
Fix issue with images not loading on first selection from media gallery

Commit:
Remove the default columns setting as we don't have access to innerBlocks at the point that the block validation is run

Commit:
Remove the default columns setting as we don't have access to innerBlocks at the point that the block validation is run

Commit:
Remove the default columns setting as we don't have access to innerBlocks at the point that the block validation is run

Commit:
Revert "Remove the default columns setting as we don't have access to innerBlocks at the point that the block validation is run"

This reverts commit 4d95e95d50ae8410868a6d7aca6b9f4c4536d36b.

Commit:
Revert "Remove the default columns setting as we don't have access to innerBlocks at the point that the block validation is run"

This reverts commit 4d95e95d50ae8410868a6d7aca6b9f4c4536d36b.

Commit:
Revert "Remove the default columns setting as we don't have access to innerBlocks at the point that the block validation is run"

This reverts commit 4d95e95d50ae8410868a6d7aca6b9f4c4536d36b.

Commit:
Add image count so we can work out default columns as innerBlocks not available at point of block validation

Commit:
Add image count so we can work out default columns as innerBlocks not available at point of block validation

Commit:
Add image count so we can work out default columns as innerBlocks not available at point of block validation

Commit:
Update Gallery to nested image blocks

Commit:
Disable the innerBlocks dropzones so drag drop works same as existing gallery - with the idea that we will improve the innerblocks drag and drop in a later PR

Commit:
Disable the innerBlocks dropzones so drag drop works same as existing gallery - with the idea that we will improve the innerblocks drag and drop in a later PR

Commit:
Lint changes

Commit:
Lint changes

dratwas committed:
WIP

Commit:
Revert "Lint changes"

This reverts commit 9b0abccdc4dc706011e315457a95418f5325f5c1.

Commit:
Revert "Lint changes"

This reverts commit 9b0abccdc4dc706011e315457a95418f5325f5c1.

Commit:
Revert "Disable the innerBlocks dropzones so drag drop works same as existing gallery - with the idea that we will improve the innerblocks drag and drop in a later PR"

This reverts commit cc05d60627d4c4970afef73306970762f39c4785.

Commit:
Revert "Disable the innerBlocks dropzones so drag drop works same as existing gallery - with the idea that we will improve the innerblocks drag and drop in a later PR"

This reverts commit cc05d60627d4c4970afef73306970762f39c4785.

Commit:
Suggested solution for handling multiple file drop into gallery

Commit:
Suggested solution for handling multiple file drop into gallery

Commit:
Remove non image files from drag and drop and disable individual image drop zone

Commit:
Remove non image files from drag and drop and disable individual image drop zone

Commit:
Fix transform to individual images

Commit:
Fix transform to individual images

Commit:
Fix transform from individual images

Commit:
Fix transform from individual images

dratwas committed:
fix dirty-image-options

dratwas committed:
fix spacing

Commit:
Revert drag and drop transform changes

Commit:
Revert drag and drop transform changes

Commit:
Add gallery transform to image block to override the default gallery transform when dragging multiple images onto the gallery

Commit:
Add gallery transform to image block to override the default gallery transform when dragging multiple images onto the gallery

dratwas committed:
Add imageCrop, blockProps to block-list and style fixes

dratwas committed:
Temporary remove the gallery from initial-html

dratwas committed:
Merge branch 'refactor/gallery-to-nested-image-blocks' into rnmobile/refactor/gallery-to-nested-image-blocks

dratwas committed:
Add gallery caption

dratwas committed:
lint fix

dratwas commented:
@iamthomasbishop 
Answering the comment from base PR https://github.com/WordPress/gutenberg/pull/25940#issuecomment-725027248

Moved that discussion here to have it in the right place since it is related only to the mobile part.

>@dratwas Assuming you are referring to the dashed child borders, why do we need to remove them? Maybe I'm missing context, but I would expect the bordering to look/work the same way as other innerBlocks do (the screenshot you shared in your comment looks correct).

Maybe I should as in a bit different way, sorry for that. My question should be: "Do we want to fit the innerBlocks with image blocks inside to keep the previous gallery look, or do we want to have default innerBlocks look"

Previous Gallery | Default InnerBlocks
--- | ---
![Simulator Screen Shot - iPhone 11 Pro - 2020-11-13 at 12 14 25](https://user-images.githubusercontent.com/16336501/99066880-fed4ed00-25a9-11eb-91d1-119e62de9fae.png) | ![Simulator Screen Shot - iPhone 11 Pro - 2020-11-13 at 12 08 06](https://user-images.githubusercontent.com/16336501/99066894-0399a100-25aa-11eb-8658-0f5ba97da132.png)

I created the test build for you here: https://github.com/wordpress-mobile/WordPress-iOS/pull/15284#issuecomment-726367909

dratwas committed:
Use grid Component instead of numColumns

dratwas committed:
Merge branch 'refactor/gallery-to-nested-image-blocks' into rnmobile/refactor/gallery-to-nested-image-blocks

dratwas committed:
fix for Android

dratwas committed:
Fix Android

iamthomasbishop commented:
> Maybe I should as in a bit different way, sorry for that. My question should be: "Do we want to fit the innerBlocks with image blocks inside to keep the previous gallery look, or do we want to have default innerBlocks look"

@dratwas Ahh okay, makes sense! My preference would be to utilize the standard innerBlocks UI (with the dashed child borders), so the second screenshot on your [last comment](https://github.com/WordPress/gutenberg/pull/26823#issuecomment-726709097) ("default innerBlocks") looks correct. 

Side note: I was trying to test the build linked in wordpress-mobile/WordPress-iOS#15284 and I'm not seeing the innerBlocks style — maybe it's behind a flag?

dratwas committed:
Use grid in list item and fix adding of new images

dratwas commented:
Thanks for the review @iamthomasbishop. I created a new build with the latest changes - https://github.com/wordpress-mobile/WordPress-iOS/pull/15284#issuecomment-726367909

Please test it and leave your thoughts here :)

Side note: I feel like after change from the old implementation (Array.map) to the innerBlocks (FlatList) the performance is noticeably worse. I can observe that when changing the count of images in a row when there are more than 5 images in the gallery.

Commit:
Move handling of file uploads to Gallery from media placeholder

Commit:
Move handling of file uploads to Gallery from media placeholder

Commit:
split innerblocks mapping into separate effect to reduce chatter

Commit:
split innerblocks mapping into separate effect to reduce chatter

dratwas committed:
Merge branch 'refactor/gallery-to-nested-image-blocks' into rnmobile/refactor/gallery-to-nested-image-blocks

dratwas committed:
Use Picker for dirt image options and add some fixes

Commit:
Add useMemo to currentImageOptions

Commit:
Add useMemo to currentImageOptions

Commit:
reuse existing innerBlocks rather than recreating with every new image selection

Commit:
reuse existing innerBlocks rather than recreating with every new image selection

Commit:
Switch to useMemo for updating local image const instead of local component state

Commit:
Switch to useMemo for updating local image const instead of local component state

Commit:
Fix issue with image sizing not being available on initial load of component some times

Commit:
Fix issue with image sizing not being available on initial load of component some times

dratwas commented:
Hey, @iamthomasbishop I changed the inline cells with `Apply to all images` and `Cancel` to the ActionSheet. Could you please take a look at the newest build? :) https://github.com/wordpress-mobile/WordPress-iOS/pull/15284#issuecomment-726367909

Commit:
Memoise the useImageSizes hook

Commit:
Memoise the useImageSizes hook

Commit:
Fix issue with media browser  defaulting to edit gallery view

Commit:
Fix issue with media browser  defaulting to edit gallery view

Commit:
Fix missed incorrect use of addToGallery

Commit:
Fix missed incorrect use of addToGallery

iamthomasbishop commented:
@dratwas That definitely feels better! Which component are we using for this on Android? I'm guessing BottomSheet?

One request: can we change the title/header label ("Gallery block options") to a more specific message/alert text like "Apply to all images in this gallery?". Here's an example (please forgive the quick dirty mock-up that I made on my phone 😅):

![image](https://user-images.githubusercontent.com/1200432/99484025-22d75c00-2925-11eb-9071-1474eabb02ae.jpeg)

One thing that seems like a bug: I'm also seeing an odd looping issue with when I first tap on the block settings icon (but if I kill the app and resume, the issue goes away). Here's a video: https://cloudup.com/cHML3GFceui . Note: it's hard to tell but after I opened the settings sheet, the alert/actionSheet popped up and continued reopening immediately upon closing it.

dratwas committed:
fix initial Picker

dratwas committed:
Change the header text of picker

dratwas committed:
Merge branch 'refactor/gallery-to-nested-image-blocks' into rnmobile/refactor/gallery-to-nested-image-blocks

dratwas committed:
fix android dirty images options

Commit:
Add some extra effects for getting the imageData as the getMedia call is async so need to keep circling through the innerblocks updates until we have all the data we need

Commit:
Add some extra effects for getting the imageData as the getMedia call is async so need to keep circling through the innerblocks updates until we have all the data we need

Commit:
Simplify the imageData by using a useSelect

Commit:
Simplify the imageData by using a useSelect

Commit:
Another optimisation - only return a new imageData reference if all images have data resolved

Commit:
Another optimisation - only return a new imageData reference if all images have data resolved

dratwas commented:
Hey, @iamthomasbishop I created new builds for iOS and Android :) Could you take a look?

I fixed the loop and changed the title of the picker.

https://github.com/wordpress-mobile/WordPress-Android/pull/13423#issuecomment-730258624
https://github.com/wordpress-mobile/WordPress-iOS/pull/15284#issuecomment-726367909

iamthomasbishop commented:
@dratwas Much much better, thank you! I have one question and it's a minor thing, not a big blocker. I'm curious if we can "pause" changes to the bottomSheet (setting label and sheet height) until _after_ the selection is made on the actionSheet? 

For example, let's say I have the `Link to` setting set to `None`. If I tap on the `Link to` cell and select another option (like `Attachment Page`) from the initial actionSheet, the new option is applied immediately (visible on the sheet) and the height of the sheet changes — _before_ the confirmation step. 

If possible, I'd like to "pause" applying the change until the user has tapped "Apply to all images" on the confirmation sheet. Alternatively, if they tap "Cancel", there should be no visible change.

aaronrobertshaw committed:
Refactored Gallery: Add loading state to gallery image size options (#27087)

* Add loading spinner for image size options
Co-authored-by: Glen Davies <glen.davies@a8c.com>

aaronrobertshaw committed:
Refactored Gallery: Add loading state to gallery image size options (#27087)

* Add loading spinner for image size options
Co-authored-by: Glen Davies <glen.davies@a8c.com>

Commit:
Initial deprecations commit

Commit:
Fix issue with linkDestination not being applied in migration

Commit:
Fix issue with linkDestination not being applied in migration

aaronrobertshaw committed:
Refactor gallery deprecations

aaronrobertshaw committed:
Refactor gallery deprecations

aaronrobertshaw committed:
Fix missing attributes from migration

aaronrobertshaw committed:
Fix missing attributes from migration

aaronrobertshaw committed:
Update deprecation to set allowResize

The imageEdit component defaults the context value for allowResize to true. The refactored gallery sets this to false by default. Setting allowResize to false when migrating a deprecated gallery allows images to be cropped in the display the same as the gallery when the post is saved and gallery reloaded.

aaronrobertshaw committed:
Update deprecation to set allowResize

The imageEdit component defaults the context value for allowResize to true. The refactored gallery sets this to false by default. Setting allowResize to false when migrating a deprecated gallery allows images to be cropped in the display the same as the gallery when the post is saved and gallery reloaded.

Commit:
Fix issue with crop not working when certain plugins are loaded

Commit:
Fix issue with crop not working when certain plugins are loaded

aaronrobertshaw committed:
Fix SCSS lint errors

aaronrobertshaw committed:
Fix SCSS lint errors

Commit:
Update the block example

Commit:
Update the block example

Commit:
Linting fixes

Commit:
Linting fixes

Commit:
Fix the e2e test and the accessibility issue with having aria group role on a list item

Commit:
Fix the e2e test and the accessibility issue with having aria group role on a list item

Commit:
fix linting issues

Commit:
fix linting issues

dratwas commented:
Hey @iamthomasbishop 
> I'm curious if we can "pause" changes to the bottomSheet (setting label and sheet height) until after the selection is made on the actionSheet?

Unfortunately in the current implementation - NO :(

This is also a shared code between the web and mobile. It sets attributes immediately and then if the user clicks "cancel" it reverts them. When the user clicks "Apply" then those changes are applied to all images inside. 

If it's not a blocker I would go with the current implementation and create an issue for that. This PR is only a mobile part of the gallery refactoring and I'm afraid that it's already too big. What do you think?

aaronrobertshaw committed:
Fix frontend omission of wp-block-image class

Also tweaks the gallery styles to remove the margins the use of `wp-block-image` introduces when it is moved to the inner `figure` element.

aaronrobertshaw committed:
Fix frontend omission of wp-block-image class

Also tweaks the gallery styles to remove the margins the use of `wp-block-image` introduces when it is moved to the inner `figure` element.

lukewalczak commented on packages/block-editor/src/components/media-placeholder/index.native.js:
```diff
@@ -163,7 +144,9 @@ function MediaPlaceholder( props ) {
 							accessibilityHint={ accessibilityHint }
 							onPress={ ( event ) => {
-								props.onFocus( event );
+								if ( props.onFocus ) {
```
Maybe add `onFocus` to props destructuring above 🤔

lukewalczak commented on packages/block-library/src/gallery/dirty-image-options.android.js:
```diff
@@ -0,0 +1,63 @@
+
+	useEffect( () => {
+		InteractionManager.runAfterInteractions( () => {
+			if ( isVisible && pickerRef.current ) {
```
```suggestion
			if ( isVisible && pickerRef?.current ) {
```

lukewalczak commented on packages/block-library/src/gallery/dirty-image-options.ios.js:
```diff
@@ -0,0 +1,45 @@
+			title={ __( 'Apply to all images in this gallery?' ) }
+		/>
+	);
+};
```
`dirty-image-options.ios.js` and `dirty-image-options.android.js` look alike, can we somehow unify both files?

lukewalczak commented on packages/block-library/src/gallery/gallery-styles.native.scss:
```diff
@@ -2,6 +2,10 @@
 }
 
+.galleryAppender {
+	padding-top: 16;
```
Use grid variable

dratwas committed:
apply CR suggestions

dratwas committed:
Merge branch 'refactor/gallery-to-nested-image-blocks' into rnmobile/refactor/gallery-to-nested-image-blocks

iamthomasbishop commented:
> If it's not a blocker I would go with the current implementation and create an issue for that.

That's totally fine, thanks @dratwas!

Commit:
change deprecation to use imageCount as isEligible check as it seems that some previous block versions may not have an ids attribute

Commit:
change deprecation to use imageCount as isEligible check as it seems that some previous block versions may not have an ids attribute

dratwas committed:
Merge branch 'refactor/gallery-to-nested-image-blocks' into rnmobile/refactor/gallery-to-nested-image-blocks

dratwas committed:
fix overflow

dratwas committed:
fix dirty-image-options

Commit:
Move back to a single deprecations file and reorder in array

Commit:
Remove additional check in v5 isEligible

Commit:
Fix the v4 migration

aaronrobertshaw committed:
Fix styles for Safari compatibility

aaronrobertshaw committed:
Remove unnecessary gallery editor styles

aaronrobertshaw committed:
Fix typo in deprecations

aaronrobertshaw committed:
Restore styles to render deprecated gallery versions

The decision to add entire deprecated.scss file was made to avoid accidental breakage on future iterations of the gallery and assist reasoning around styles. These outweighed the minor benefit in reducing a few style declarations.

This aalso contains some minor tweaks to fix inconsistencies and clean up dead styles.

aaronrobertshaw committed:
Avoid applying flex styles to IE11

Commit:
Add additional selector to prevent the hidden individual image drop zones ending up display flex instead of hidden

aaronrobertshaw committed:
IE11 styling improvements

Commit:
Apply default style class to new images added to gallery

Commit:
fix linting issues

Commit:
Move block props to the outer wrapper

Commit:
Revert "Move block props to the outer wrapper"

This reverts commit e0723a7068cd56cb83bb80ded67444d7676aa4f8.

Commit:
Revert "Revert "Move block props to the outer wrapper""

This reverts commit a7504fd8386efd46577e2d9a566e7b816404e090.

Commit:
put media uploader outside figure so structure matches front end

Commit:
Replace div with View for the sake of native code

glendaviesnz committed:
Gallery Block Refactor: Account for null image ids in gallery migrations (#27855)

Co-authored-by: Glen Davies <glen.davies@a8c.com>

glendaviesnz committed:
Remove the gradient and put caption under image if is-rounded style applied (#27869)

* Add alignment fixes for non cropped images

Co-authored-by: Glen Davies <glen.davies@a8c.com>

Commit:
Update Gallery to nested image blocks

Commit:
Initial deprecations commit

Commit:
Fetch media if isListItem is true

Commit:
Change context from isListItem to isGrouped

Commit:
Remove external div wrapper by moving InnerBlocks to a fragment

Commit:
remove extra wrapper around media placeholder and caption and use flex css instead

Commit:
Revert "remove extra wrapper around media placeholder and caption and use flex css instead

Commit:
Revert "Remove external div wrapper by moving InnerBlocks to a fragment"

Commit:
another update to image wrapper

Commit:
Move setting of attributes to the child images

Commit:
Remove outer div wrapper

aaronrobertshaw committed:
Keep image margins while dragging sibling

Commit:
Fix e2e test expected markup to match new structure

dratwas committed:
merge with base branch

dratwas committed:
resolve conflicts

dratwas committed:
remove unused deprecation files

dratwas committed:
revert changes in picker

dratwas committed:
fix web scss changes after rebase

dratwas committed:
remove not needed change

dratwas committed:
remove dirty image options

dratwas committed:
rever changes in gallery edit

dratwas committed:
WIP use context to set target destination and slug

dratwas committed:
Fix inherted image props in gallery

dratwas committed:
fix key in bloc-styles

chipsnyder commented:
Closing this PR in favor of the updated mobile PR https://github.com/WordPress/gutenberg/pull/31306
